cpmpy/transformations/linearize.py

```diff
@@ -63,18 +70,142 @@
+
+def linearize_mul_comparison(cpm_expr: Comparison, supported: Set[str] = {}, reified: bool = False, csemap: Optional[Dict[Any, Any]] = None) -> List[Expression]:
+    """
+    Linearizes multiplication comparisons (const*var | bool*bool | bool*int | int*int) <cmp> rhs.
 
-def linearize_constraint(lst_of_expr, supported={"sum","wsum","->"}, reified=False, csemap=None):
+    Arguments:
+        cpm_expr (Comparison): Comparison to linearize, e.g. `bool*bool <cmp> rhs` or `bool*int <cmp> rhs` or `int*int <cmp> rhs`.
+        supported (Set[str]): Set of constraint and variable type names that are supported by the target
+            solver, e.g. `{"sum", "wsum", "->", "and", "or", "alldifferent"}`.
+        reified (bool): Whether the constraint is fully reified. When True, nested implications are linearized using Big-M method instead of indicator constraints.
+        csemap (Optional[Dict[Any, Any]]): Optional dictionary for common subexpression elimination, mapping expressions to their flattened variable representations. Used to avoid creating duplicate variables for the same subexpression.
+
+    Returns:
+        List of linearized CPMpy expressions.
+    """
+    lhs, rhs = cpm_expr.args
+    assert lhs.name == "mul", f"linearize_mul_comparison expects a multiplication comparison, got {lhs}"
+
+    if is_num(lhs.args[0]): # const * iv <comp> rhs
+        lhs = Operator("wsum",[[lhs.args[0]], [lhs.args[1]]])
+        return linearize_constraint([eval_comparison(cpm_expr.name, lhs, rhs)], supported=supported, reified=reified, csemap=csemap)
+    
+    bv_idx = None
+    if isinstance(lhs.args[0], _BoolVarImpl):
+        bv_idx = 0
+    elif isinstance(lhs.args[1], _BoolVarImpl):
+        bv_idx = 1
+
+    if bv_idx is not None:
+        # bv * iv <comp> rhs, rewrite to (bv -> iv <comp> rhs) & (~bv -> 0 <comp> rhs)
+        # (also works for bv1*bv2 <comp> rhs)
+        bv, iv = lhs.args[bv_idx], lhs.args[1-bv_idx]
+        bv_true = bv.implies(eval_comparison(cpm_expr.name, iv, rhs))
+        bv_false = (~bv).implies(eval_comparison(cpm_expr.name, 0, rhs))
+        return linearize_constraint(simplify_boolean([bv_true, bv_false]), supported=supported, reified=reified, csemap=csemap)
+    else:
+        # iv1 * iv2 <comp> rhs, do Boolean expansion of smallest integer
+        # resulting in sum([v*b*i2 for v,b in int2bool(i1)]) <comp> rhs
+
+        # choose smallest integer as i1 (to minimize the number of Boolean variables)
+        i1,i2 = lhs.args
+        leni1 = (i1.ub - i1.lb) + 1
+        leni2 = (i2.ub - i2.lb) + 1
+        if leni2 < leni1:
+            # swap, call i1 the smallest integer
+            i1,i2 = i2,i1
+
+        # encode i1 with Booleans (with temproary ivarmap)
+        encoding = "direct"
+        if min(leni1,leni2) >= 8:  # arbitrary heuristic
+            encoding = "binary"  # results in fewer Bools
+        i1_enc, cons = _encode_int_var({}, i1, encoding)  # {}: no ivarmap used
```

Comment:
So here there is no csemap-check?
That means that if we linearize two constraints (x * y <= 4) and (x * z <= 6); we will Booleanize x twice?